lib/khttp/ktest/mocks.go

```diff
@@ -0,0 +1,145 @@
+// +build !release
+
+package ktest
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	rpcAstore "github.com/enfabrica/enkit/astore/rpc/astore"
+	"github.com/enfabrica/enkit/astore/server/astore"
+	"github.com/enfabrica/enkit/lib/srand"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/test/bufconn"
+	"log"
+	"math/rand"
+	"net"
+	"os"
+	"os/exec"
+	"strings"
+	"syscall"
+)
+
+type MinioDescriptor struct {
+	Port uint16
+}
+
+//RunMinioServer will spin up a minio serer using the local docker daemon
+//it also returns a func to call that will close and destroy the running image
+//the port and network bind are determined by docker and returned
+func RunMinioServer() (MinioDescriptor, func() error, error) {
+	_ = "gcloud beta emulators datastore start --no-store-on-disk"
+	return MinioDescriptor{}, nil, nil
+}
+
+type EmulatedDatastoreDescriptor struct {
+	Addr *net.TCPAddr
+}
+
+func RunEmulatedDatastore() (*EmulatedDatastoreDescriptor, KillAbleProcess, error) {
+	emulatorAddr, err := AllocatePort()
+	if err != nil {
+		return nil, nil, err
+	}
+	cmd := exec.Command("gcloud",
+		"beta", "emulators", "datastore", "start",
+		"--no-store-on-disk",
+		fmt.Sprintf("--host-port=127.0.0.1:%d", emulatorAddr.Port),
+		"--quiet")
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+	outputStdErrPipe, err := cmd.StderrPipe()
+	if err != nil {
+		return nil, nil, err
+	}
+	err = cmd.Start()
+	killFunc := []func(){
+		func() {
+			fmt.Println("killing emulator datastore")
+			if err := cmd.Process.Kill(); err != nil {
+				log.Fatalln(fmt.Sprintf("error killing process %d", cmd.Process.Pid))
+			}
+		}}
+	if err != nil {
+		return nil, killFunc, err
+	}
+	datastoreBooted := make(chan bool)
+	//TODO concatenate stdout and stderr?
+	//the datastore emulator writes all logs to the error channel for some reason
+	scannerErr := bufio.NewScanner(outputStdErrPipe)
+	//TODO add timeout on starting, maybe 3 seconds?
+	go func() {
+		for scannerErr.Scan() {
+			if strings.Contains(scannerErr.Text(), "Dev App Server is now running") {
+				datastoreBooted <- true
+			}
+		}
+	}()
+	//TODO pass failure logs to errors.Wrap upstream
+	if !<-datastoreBooted {
+		return nil, killFunc, errors.New("unable to start the datastore simulator for reason Y")
+	} else {
+		return &EmulatedDatastoreDescriptor{
+			Addr: emulatorAddr,
+		}, killFunc, nil
+	}
+}
+
+type AStoreDescriptor struct {
+	Connection *grpc.ClientConn
+	Server     *astore.Server
+}
+
+//Run Astoreserver will spin up an emulated datastore along with an instance of the astore grpc server
+func RunAStoreServer() (*AStoreDescriptor, KillAbleProcess, error) {
+
+	killFunctions := KillAbleProcess([]func(){})
+	emulatorDescriptor, emulatorKill, err := RunEmulatedDatastore()
+	killFunctions.AddKillable(emulatorKill)
+	if err != nil {
+		return nil, killFunctions, err
+	}
+	err = os.Setenv(
+		"STORAGE_EMULATOR_HOST",
```

Comment:
who / what needs this environment variable? What is the effect of this variable? Can you add a comment? Like:

```
// Google library github.com/google/whatever/storage looks at a STORAGE_EMULATOR_HOST before connecting.
// When set, it causes the server to ignore ... configuration parameters, and ...
```

(reason I'm asking: certainly don't remember writing code looking at this env variable, I have no idea what the effects are)

Reply:
https://github.com/googleapis/google-cloud-go/blob/86178128cae668f46d014569d5c25844aa379cdb/storage/storage.go#L116

All google go client (well most) take in some version that environment variable which lets it redirect from google's official endpoints. Unfortunately you can only configure it though os.Env and generally not through opts.

Reply:
makes sense. Minor nit: would add a // Causes the google-could-go/storage library to use a local emulator rather than the real endpoint. See https://github.com/googleapis/google-cloud-go/blob/86178128cae668f46d014569d5c25844aa379cdb/storage/storage.go#L109.